<a href="https://colab.research.google.com/github/jcardonamde/food-sales-predictions/blob/main/food_sales_predictions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Parte 1**

Importar libreria y dataset

In [46]:
import pandas as pd

In [47]:
file_food_sales = '/content/drive/MyDrive/Colab Notebooks/Codificación de Dojo/Proyecto/sales_predictions.csv'
df = pd.read_csv(file_food_sales)
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

# **Parte 2**

El primer proyecto será de una predicción de ventas para productos alimenticios vendidos en diversas tiendas.  El objetivo de esto es ayudar al minorista a comprender las propiedades de los productos y los puntos de venta que desempeñan un papel crucial en el aumento de las ventas.

Para la parte 2, la tarea es usar Pandas para empezar a limpiar y explorar los datos. Al ser científicos de datos, alrededor del 80% de nuestro tiempo se dedica a la preparación de datos, así que este es un paso muy importante. 



*   Sus datos ya deberían estar cargados en el cuaderno de la parte 1 del proyecto.
*   Exploren los datos - ¿Qué necesitan para limpiar estos datos? ¡Límpienlos!
*   Asegúrense de abordar cada una de las siguientes tareas:



  


**1)** ¿Cuántas filas y columnas?

In [48]:
df.shape

(8523, 12)

Tenemos en nuestro dataset 8523 filas y 12 columnas.

**2)** ¿Cuáles son los tipos de datos en cada variable?

In [49]:
# Mediante la función de info obtenemos el tipo de dato en cada columna del dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


**3)** ¿Hay duplicados? Si es el caso, eliminen algunos duplicados.

In [50]:
# Identificamos cual es la cantidad de datos duplicados
df.duplicated()
df.duplicated().sum()

0

Al momento no se tienen datos duplicados en el dataset



---



**4)** Identifiquen los valores faltantes.

Revisando la información que me trajo el anterior paso 2. Pude identificar que las columnas de Item_Weight y Outlet_Size estan por debajo del valor de datos homogeneo que encontramos en el resto de columnas, que en este caso seria de 8523 valores.

In [51]:
# Inicialmente examino el número de valores unicos entre las columnas con las cuales estamos trabajando
df.nunique()

Item_Identifier              1559
Item_Weight                   415
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     3
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

Con la anterior función puedo identificar que Item_Weight tiene 415 datos unicos y Outlet_Size 3 datos unicos.

In [52]:
# Luego verificamos cuanto es el número de filas sin información en estas columnas
df.isna().sum()

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Nos encontramos que tenemos 1463 datos faltantes en la columna Item_Weight y tenemos 2410 datos faltantes en la columna Outlet_Size



---



**5)** Decidan cómo abordar los valores faltantes y cómo hacerlo. (Esto requiere sus criterios, así que expliquen su elección).

In [53]:
# Hacemos un conteo de los valores de esta columna para identificar como esta compuesta
df['Item_Weight'].value_counts()

12.150    86
17.600    82
13.650    77
11.800    76
15.100    68
          ..
7.275      2
7.685      1
9.420      1
6.520      1
5.400      1
Name: Item_Weight, Length: 415, dtype: int64

In [54]:
# Realizo un filtrado por los valores faltantes de la columna Item_Weight para validar en el dataframe como se ven las columnas 
item_weight_missing = df['Item_Weight'].isna()
df.loc[item_weight_missing,:]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8485           DRK37          NaN          Low Fat         0.043792   
8487           DRG13          NaN          Low Fat         0.037006   
8488           NCN14          NaN          Low Fat         0.091473   
8490           FDU44          NaN          Regular         0.102296   
8504           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

Me llamo la atención que habia cierta relación de los datos faltantes con las columnas Outlet_Identifier, Outlet_Establishment_Year y Item_Fat_Content. Procedo a hacer una analisis de agrupar por cada columna junto con la sumatoria de los valores para identificar patrones.

In [55]:
# Hago un agrupamiento y sumatoria por la columna de Outlet_Identifier para analizar 
df.groupby(['Outlet_Identifier']).sum()

Item_Weight  Item_Visibility     Item_MRP  \
Outlet_Identifier                                              
OUT010                7166.800        56.308832   78131.5646   
OUT013               12121.730        55.879859  131809.0156   
OUT017               11877.495        56.834646  129103.9564   
OUT018               11946.465        56.621454  131477.7724   
OUT019                   0.000        57.257041   73807.5824   
OUT027                   0.000        54.804764  130714.6746   
OUT035               11931.295        56.974873  133103.9070   
OUT045               11751.840        56.180780  130942.7782   
OUT046               11966.125        56.231875  132113.3698   
OUT049               12013.225        56.549156  130476.8598   

                   Outlet_Establishment_Year  Item_Outlet_Sales  
Outlet_Identifier                                                
OUT010                               1108890       1.883402e+05  
OUT013                               1851884       2.142664e+06  
OUT017                               1858482       2.167465e+06  
OUT018                               1864352       1.851823e+06  
OUT019                               1048080       1.796941e+05  
OUT027                               1855975       3.453926e+06  
OUT035                               1863720       2.268123e+06  
OUT045                               1859858       2.036725e+06  
OUT046                               1857210       2.118395e+06  
OUT049                               1859070       2.183970e+06

Me encuentro de que no hay valores de Item_Weight para dos de los grupos de la Columna Outlet_Identifier. Asi que sigo analizando la siguiente columna.

In [56]:
# Hago un agrupamiento y sumatoria por la columna de Outlet_Establishment_Yearr para analizar 
df.groupby(['Outlet_Establishment_Year']).sum()

Item_Weight  Item_Visibility     Item_MRP  \
Outlet_Establishment_Year                                              
1985                             0.000       112.061805  204522.2570   
1987                         12121.730        55.879859  131809.0156   
1997                         11966.125        56.231875  132113.3698   
1998                          7166.800        56.308832   78131.5646   
1999                         12013.225        56.549156  130476.8598   
2002                         11751.840        56.180780  130942.7782   
2004                         11931.295        56.974873  133103.9070   
2007                         11877.495        56.834646  129103.9564   
2009                         11946.465        56.621454  131477.7724   

                           Item_Outlet_Sales  
Outlet_Establishment_Year                     
1985                            3.633620e+06  
1987                            2.142664e+06  
1997                            2.118395e+06  
1998                            1.883402e+05  
1999                            2.183970e+06  
2002                            2.036725e+06  
2004                            2.268123e+06  
2007                            2.167465e+06  
2009                            1.851823e+06

En este caso los valores para uno de los grupos de la Columna Outlet_Establishment_Year con Item_Weight no existen. Asi que sigo analizando la siguiente columna.

In [57]:
df.groupby(['Item_Fat_Content']).sum()

Item_Weight  Item_Visibility     Item_MRP  \
Item_Fat_Content                                              
LF                   3328.835        21.054330   43857.1062   
Low Fat             54687.900       326.541080  717390.8394   
Regular             30341.545       200.970026  409413.1214   
low fat              1055.375         7.313655   15071.7328   
reg                  1361.320         7.764190   15948.6810   

                  Outlet_Establishment_Year  Item_Outlet_Sales  
Item_Fat_Content                                                
LF                                   631348       6.552424e+05  
Low Fat                            10167044       1.101503e+07  
Regular                             5771667       6.457454e+06  
low fat                              223539       2.338270e+05  
reg                                  233923       2.295765e+05

En este caso me encuentro que agrupando la columna Item_Fat_Content me trae valores para la relación con la columna Item_Weight. Desde allí podria intuir que sacando una media de estos podria asignarle a la columna que tiene los valores faltantes, dichos valores con el promedio que se pueda obtener.

In [58]:
# Usamos la opción de la media para el relleno de la columna Item_Weight en vista de que el datatype de dicha columna es de tipo Float
median_Item_Weight = df['Item_Weight'].mean()
df['Item_Weight'].fillna(median_Item_Weight, inplace=True)
df.isna().sum()

Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Como vemos ya solo tenemos la columna Outlet_Size aun con valores faltantes, y la de Item_Weight rellenada sus anteriores campos vacios con los nuevos datos de la media

In [59]:
# Realizo un filtrado por los valores faltantes de la columna Outlet_Size para validar en el dataframe como se ven las columnas 
Outlet_Size_missing = df['Outlet_Size'].isna()
df.loc[Outlet_Size_missing,:]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07       19.200          Regular         0.000000   
8              FDH17       16.200          Regular         0.016687   
9              FDU28       19.200          Regular         0.094450   
25             NCD06       13.000          Low Fat         0.099887   
28             FDE51        5.925          Regular         0.161467   
...              ...          ...              ...              ...   
8502           NCH43        8.420          Low Fat         0.070712   
8508           FDW31       11.350          Regular         0.043246   
8509           FDG45        8.100          Low Fat         0.214306   
8514           FDA01       15.000          Regular         0.054489   
8519           FDS36        8.380          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
28                    Dairy   45.5086            OUT010   
...                     ...       ...               ...   
8502              Household  216.4192            OUT045   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
25                         2007         NaN               Tier 2   
28                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8502                       2002         NaN               Tier 2   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

Deseo validar si hay alguna relación de datos faltantes con cada una de estas columnas Outlet_Establishment_Year, Outlet_Size	Outlet_Location_Type, Outlet_Type para el caso de la columna Outlet_Size. 

In [60]:
# Hacemos un conteo de los valores de esta columna para identificar como esta compuesta
df.groupby(['Outlet_Type']).sum()

Item_Weight  Item_Visibility     Item_MRP  \
Outlet_Type                                                     
Grocery Store      13955.636657       113.565873  151939.1470   
Supermarket Type1  71661.710000       338.651189  787549.8868   
Supermarket Type2  11946.465000        56.621454  131477.7724   
Supermarket Type3  12021.898247        54.804764  130714.6746   

                   Outlet_Establishment_Year  Item_Outlet_Sales  
Outlet_Type                                                      
Grocery Store                        2156970       3.680343e+05  
Supermarket Type1                   11150224       1.291734e+07  
Supermarket Type2                    1864352       1.851823e+06  
Supermarket Type3                    1855975       3.453926e+06

In [61]:
# Hacemos un conteo de los valores de esta columna para identificar como esta compuesta
df['Outlet_Establishment_Year'].value_counts()

1985    1463
1987     932
1999     930
1997     930
2004     930
2002     929
2009     928
2007     926
1998     555
Name: Outlet_Establishment_Year, dtype: int64

In [62]:
# Hacemos un conteo de los valores de esta columna para identificar como esta compuesta
df['Outlet_Location_Type'].value_counts()

Tier 3    3350
Tier 2    2785
Tier 1    2388
Name: Outlet_Location_Type, dtype: int64

In [63]:
most_common_Outlet_Size = df['Outlet_Size'].mode()
most_common_Outlet_Size

0    Medium
dtype: object

Como no he visto algún patron de relación notable y he definido que esta columna no vaya a influir en primaria instancia en posteriores analisis. Por lo que se creara una nueva 

In [64]:
# Se llenan los datos faltantes con una nueva categoria
df['Outlet_Size'].fillna('Missing', inplace=True)



---



**6)** Confirmen que no hay valores faltantes después de abordarlos.

In [65]:
# Verifico nuevamente con la función isna si tenemos datos faltantes luego de intervenir el dataset inicial
df.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64



---



**7)** Encuentren y arreglen alguna categoría inconsistente de datos (example: fix cat, Cat, and cats so that they are consistent) 

In [66]:
df.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type', 'Item_Outlet_Sales'],
      dtype='object')

In [67]:
# Haciendo una revisión columna por columna nos encontramos que la columna Item_Fat_Content tiene algunos errores en sus valores categoricos
# Se procede a realizar el respectivo correctivo para relacionar los valores errores hacia las categorias Low Fat y Regular
df['Item_Fat_Content'].value_counts()

Low Fat    5089
Regular    2889
LF          316
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [68]:
# Usando el método replace para hacer la correción en conversión de un valor categorico en especial
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('LF', 'Low Fat')
df['Item_Fat_Content'].value_counts()

Low Fat    5405
Regular    2889
reg         117
low fat     112
Name: Item_Fat_Content, dtype: int64

In [69]:
# Usando el método replace para hacer la correción en conversión de un valor categorico en especial
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('low fat', 'Low Fat')
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    2889
reg         117
Name: Item_Fat_Content, dtype: int64

In [70]:
# Usando el método replace para hacer la correción en conversión de un valor categorico en especial
df['Item_Fat_Content'] = df['Item_Fat_Content'].replace('reg', 'Regular')
df['Item_Fat_Content'].value_counts()

Low Fat    5517
Regular    3006
Name: Item_Fat_Content, dtype: int64

In [71]:
# En este caso encontramos un caso especial es que vemos que hay unos 526 items sin visibilidad, podria ser un dato correcto como quizas un posible error
# Por ahora lo dejaremos en observación, pero ya lo tenemos identificado
df['Item_Visibility'].value_counts()

0.000000    526
0.076975      3
0.162462      2
0.076841      2
0.073562      2
           ... 
0.013957      1
0.110460      1
0.124646      1
0.054142      1
0.044878      1
Name: Item_Visibility, Length: 7880, dtype: int64



---



8) Para cualquier columna numérica, obtengan las estadísticas resumidas para cada uno (mínimo, máximo y media)

In [72]:
# Mostramos con la función describe todas las estadísticas de las columnas numericas
df.describe()

Item_Weight  Item_Visibility     Item_MRP  Outlet_Establishment_Year  \
count  8523.000000      8523.000000  8523.000000                8523.000000   
mean     12.857645         0.066132   140.992782                1997.831867   
std       4.226124         0.051598    62.275067                   8.371760   
min       4.555000         0.000000    31.290000                1985.000000   
25%       9.310000         0.026989    93.826500                1987.000000   
50%      12.857645         0.053931   143.012800                1999.000000   
75%      16.000000         0.094585   185.643700                2004.000000   
max      21.350000         0.328391   266.888400                2009.000000   

       Item_Outlet_Sales  
count        8523.000000  
mean         2181.288914  
std          1706.499616  
min            33.290000  
25%           834.247400  
50%          1794.331000  
75%          3101.296400  
max         13086.964800